### PR TITLE
Initial work on experiment specific report files

### PR DESCRIPTION
**Pull request by vulder**
Depends on PolyJIT/benchbuild#441 otherwise, we cannot correctly pass on the args to __init_subclass__

### CODE REVIEW

File: varats-core/varats/report/report.py [hunk 1]
```diff
@@ -252,17 +253,30 @@ def commit_hash(self) -> ShortCommitHash:
         raise ValueError(f'File {self.filename} name was wrongly formated.')
 
     @property
-    def shorthand(self) -> str:
+    def experiment_shorthand(self) -> str:
+        """
+        Experiment shorthand of the result file.
+
+        Returns:
+            the experiment  shorthand from a result file
```
Comment (boehmseb) [inline varats-core/varats/report/report.py]:
```suggestion
            the experiment shorthand from a result file
```

File: varats-core/varats/report/report.py [hunk 2]
```diff
@@ -351,13 +368,17 @@ def __init__(self, path: Path) -> None:
...
     @classmethod
-    def __init_subclass__(cls, *args: tp.Any, **kwargs: tp.Any) -> None:
+    def __init_subclass__(
+        cls, shorthand: str, file_type: str, *args: tp.Any, **kwargs: tp.Any
+    ) -> None:
         # mypy does not yet fully understand __init_subclass__()
         # https://github.com/python/mypy/issues/4660
         super().__init_subclass__(*args, **kwargs)  # type: ignore
 
+        cls.SHORTHAND = shorthand
```
Comment (boehmseb) [inline varats-core/varats/report/report.py:383]: mypy is right here. You need to add class-level declarations.

File: varats-core/varats/experiment/experiment_util.py [hunk 3]
```diff
@@ -244,6 +249,21 @@ class VersionExperiment(Experiment):  # type: ignore
...
+        # mypy does not yet fully understand __init_subclass__()
+        # https://github.com/python/mypy/issues/4660
+        super().__init_subclass__(*args, **kwargs)  # type: ignore
+
+        cls.SHORTHAND = shorthand
+
+    @staticmethod
+    @abstractmethod
+    def shorthand() -> str:
+        """Experiment shorthand."""
```
Comment (boehmseb) [inline varats-core/varats/experiment/experiment_util.py]: I guess this should work the same as with reports:

```suggestion
    @classmethod
    def shorthand(cls) -> str:
        """Experiment shorthand."""
        return cls.SHORTHAND
```

**Review by boehmseb — CHANGES_REQUESTED** [review]
I have not much to add from a conceptional point of view since we already discussed this in person. But I found a few smaller things, basically the usual stuff :wink:

File: varats-core/varats/report/report.py (hunk 2 above)
Comment (vulder) [reply varats-core/varats/report/report.py:383]: I know, I'm still trying to fix out all the bugs before I fix type decls

File: tests/data/test_diff_correlation_overview_table.py [hunk 4]
```diff
@@ -28,4 +28,7 @@ def test_table_tex_output(self):
         ).tabulate()
 
         with open("tables/b_diff_correlation_overview.tex") as expected:
+            #print(table)
```
Comment (cormensratio) [inline tests/data/test_diff_correlation_overview_table.py]: Remove debug prints if not needed anymore.

File: varats-core/varats/report/report.py [hunk 5]
```diff
@@ -252,17 +253,30 @@ def commit_hash(self) -> ShortCommitHash:
...
+        Experiment shorthand of the result file.
+
+        Returns:
+            the experiment  shorthand from a result file
+        """
+        match = ReportFilename.__RESULT_FILE_REGEX.search(self.filename)
+        if match:
+            return match.group("experiment_shorthand")
+
+        raise ValueError(f'File {self.filename} name was wrongly formated.')
```
Comment (cormensratio) [inline varats-core/varats/report/report.py]:
```suggestion
        raise ValueError(f'File {self.filename} name was wrongly formatted.')
```

File: varats-core/varats/report/report.py [hunk 6]
```diff
@@ -319,15 +335,16 @@ def get_file_name(
         """
         status_ext = FileStatusExtension.get_status_extension(extension_type)
 
-        # Add the missing '.' if none was given by the report
+        #  Add the missing '.' if none was given by the report
```
Comment (cormensratio) [inline varats-core/varats/report/report.py]:
```suggestion
        # Add the missing '.' if none was given by the report
```

File: varats-core/varats/report/report.py [hunk 7]
```diff
@@ -319,15 +335,16 @@ def get_file_name(
...
         if file_ext and not file_ext.startswith("."):
             file_ext = "." + file_ext
 
         return ReportFilename.__RESULT_FILE_TEMPLATE.format(
```
Comment (cormensratio) [inline varats-core/varats/report/report.py]: Returns a `str` and not the required `'ReportFilename'`.
Use:
```
ReportFilename.get_file_name(
...
```

File: varats/varats/data/reports/blame_report.py [hunk 8]
```diff
@@ -329,34 +326,6 @@ def shorthand(cls) -> str:
         """Shorthand for this report."""
         return cls.SHORTHAND
```
Comment (cormensratio) [inline varats/varats/data/reports/blame_report.py]: `SHORTHAND` got replaced by `shorthand()`
```suggestion
        return cls.shorthand()
```

File: varats-core/varats/report/gnu_time_report.py [hunk 9]
```diff
@@ -6,19 +6,17 @@
...
+from varats.utils.git_util import ShortCommitHash
 from varats.utils.util import static_vars
 
 
 class WrongTimeReportFormat(Exception):
     """Thrown if the a time report could not be parsed."""
 
 
-class TimeReport(BaseReport):
+class TimeReport(BaseReport, shorthand="TR", file_type=""):
```
Comment (cormensratio) [inline varats-core/varats/report/gnu_time_report.py:16]: Should the `file_type` be an empty string?

**Review by cormensratio — CHANGES_REQUESTED** [review]
Besides my comments and mypys' the `get_file_name()` function in the experiments misses the new `experiment_shorthand` parameter. 
From a conceptional point of view this should allow us to identify reports specific to their experiments, but some test runs may not hurt.

File: varats-core/varats/report/gnu_time_report.py (hunk 9 above)
Comment (vulder) [reply varats-core/varats/report/gnu_time_report.py:16]: yes

File: tests/data/test_report.py [hunk 10]
```diff
@@ -80,44 +81,70 @@ def test_filename(self):
...
         )
-        self.assertEqual(self.report_filename.shorthand, "CR")
+        self.assertEqual(self.report_filename.report_shorthand, "CR")
+        self.assertEqual(self.report_filename.experiment_shorthand, "CRE")
         self.assertEqual(
             self.report_filename.file_status, FileStatusExtension.SUCCESS
         )
 
+    def test_accessors_broken(self):
+        """Tests if the different accessor functions work."""
```
Comment (boehmseb) [inline tests/data/test_report.py]: "work" might be a bit misleading if `ValueError`s are expected :wink:

**Review by boehmseb — APPROVED** [review]
lgtm

**Review by cormensratio — APPROVED** [review]
Looks good 👍🏽.

**Pull request merged**